// @cyberlab/cyberconnect-v2/lib/cyberConnect.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var network_1 = require("./network");
var queries_1 = require("./queries");
var error_1 = require("./error");
var types_1 = require("./types");
var utils_1 = require("./utils");
var _1 = require(".");
var constant_1 = require("./constant");
var crypto_1 = require("./crypto");
var CyberConnect = /** @class */ (function () {
    function CyberConnect(config) {
        this.address = '';
        this.signature = '';
        this.chain = types_1.Blockchain.ETH;
        this.chainRef = '';
        this.provider = null;
        this.signingMessageEntity = '';
        var provider = config.provider, namespace = config.namespace, env = config.env, chainRef = config.chainRef, chain = config.chain, signingMessageEntity = config.signingMessageEntity;
        if (!namespace) {
            throw new error_1.ConnectError(error_1.ErrorCode.EmptyNamespace);
        }
        this.namespace = namespace;
        this.endpoint = network_1.endpoints[env || _1.Env.PRODUCTION];
        this.chain = chain || types_1.Blockchain.ETH;
        this.chainRef = chainRef || '';
        this.provider = provider;
        this.signingMessageEntity = signingMessageEntity;
        delete window.localStorage[constant_1.C_ACCESS_TOKEN_KEY];
    }
    CyberConnect.prototype.getAddress = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.address) {
                            return [2 /*return*/, this.address];
                        }
                        _a = this;
                        return [4 /*yield*/, (0, utils_1.getAddressByProvider)(this.provider, this.chain)];
                    case 1: return [2 /*return*/, (_a.address = _b.sent())];
                }
            });
        });
    };
    CyberConnect.prototype.authWithSigningKey = function () {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var publicKey, acknowledgement, message, _c, signingKeySignature, resp, e_1;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, (0, crypto_1.hasSigningKey)(this.address)];
                    case 1:
                        if (_d.sent()) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, (0, crypto_1.getPublicKey)(this.address)];
                    case 2:
                        publicKey = _d.sent();
                        acknowledgement = "I authorize ".concat(this.signingMessageEntity || 'CyberConnect', " from this device using signing key:\n");
                        message = "".concat(acknowledgement).concat(publicKey);
                        _c = this;
                        return [4 /*yield*/, this.getAddress()];
                    case 3:
                        _c.address = _d.sent();
                        _d.label = 4;
                    case 4:
                        _d.trys.push([4, 9, , 10]);
                        return [4 /*yield*/, (0, utils_1.getSigningKeySignature)(this.provider, this.chain, message, this.address)];
                    case 5:
                        signingKeySignature = _d.sent();
                        if (!signingKeySignature) return [3 /*break*/, 7];
                        return [4 /*yield*/, (0, queries_1.registerSigningKey)({
                                address: this.address,
                                signature: signingKeySignature,
                                message: message,
                                url: this.endpoint.cyberConnectApi,
                            })];
                    case 6:
                        resp = _d.sent();
                        if (((_a = resp === null || resp === void 0 ? void 0 : resp.data) === null || _a === void 0 ? void 0 : _a.registerSigningKey.status) !== 'SUCCESS') {
                            throw new error_1.ConnectError(error_1.ErrorCode.GraphqlError, (_b = resp === null || resp === void 0 ? void 0 : resp.data) === null || _b === void 0 ? void 0 : _b.registerSigningKey.result);
                        }
                        return [3 /*break*/, 8];
                    case 7: throw new Error('signingKeySignature is empty');
                    case 8: return [3 /*break*/, 10];
                    case 9:
                        e_1 = _d.sent();
                        console.log('e', e_1);
                        (0, crypto_1.clearSigningKeyByAddress)(this.address);
                        throw new Error('User cancel the sign process');
                    case 10: return [2 /*return*/];
                }
            });
        });
    };
    CyberConnect.prototype.follow = function (address, handle) {
        var _a, _b, _c, _d;
        console.log('e');
        return __awaiter(this, void 0, void 0, function () {
            var _e, message, signature, publicKey, params, resp, e_2;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _f.trys.push([0, 8, , 9]);
                        _e = this;
                        return [4 /*yield*/, this.getAddress()];
                    case 1:
                        _e.address = _f.sent();
                        return [4 /*yield*/, this.authWithSigningKey()];
                    case 2:
                        _f.sent();
                        message = {
                            op: 'follow',
                            address: address,
                            handle: handle,
                            ts: Date.now(),
                        };
                        return [4 /*yield*/, (0, crypto_1.signWithSigningKey)(JSON.stringify(message), this.address)];
                    case 3:
                        signature = _f.sent();
                        return [4 /*yield*/, (0, crypto_1.getPublicKey)(this.address)];
                    case 4:
                        publicKey = _f.sent();
                        params = {
                            address: address,
                            handle: handle,
                            message: JSON.stringify(message),
                            signature: signature,
                            signingKey: publicKey,
                        };
                        return [4 /*yield*/, (0, queries_1.follow)(params, this.endpoint.cyberConnectApi)];
                    case 5:
                        resp = _f.sent();
                        if (!(((_a = resp === null || resp === void 0 ? void 0 : resp.data) === null || _a === void 0 ? void 0 : _a.follow.status) === 'INVALID_SIGNATURE')) return [3 /*break*/, 7];
                        return [4 /*yield*/, (0, crypto_1.clearSigningKey)()];
                    case 6:
                        _f.sent();
                        throw new error_1.ConnectError(error_1.ErrorCode.GraphqlError, (_b = resp === null || resp === void 0 ? void 0 : resp.data) === null || _b === void 0 ? void 0 : _b.follow.status);
                    case 7:
                        if (((_c = resp === null || resp === void 0 ? void 0 : resp.data) === null || _c === void 0 ? void 0 : _c.follow.status) !== 'SUCCESS') {
                            throw new error_1.ConnectError(error_1.ErrorCode.GraphqlError, (_d = resp === null || resp === void 0 ? void 0 : resp.data) === null || _d === void 0 ? void 0 : _d.follow.status);
                        }
                        return [3 /*break*/, 9];
                    case 8:
                        e_2 = _f.sent();
                        throw new error_1.ConnectError(error_1.ErrorCode.GraphqlError, e_2.message || e_2);
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    CyberConnect.prototype.unfollow = function (address, handle) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function () {
            var _e, message, signature, publicKey, params, resp, e_3;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _f.trys.push([0, 8, , 9]);
                        _e = this;
                        return [4 /*yield*/, this.getAddress()];
                    case 1:
                        _e.address = _f.sent();
                        return [4 /*yield*/, this.authWithSigningKey()];
                    case 2:
                        _f.sent();
                        message = {
                            op: 'unfollow',
                            address: address,
                            handle: handle,
                            ts: Date.now(),
                        };
                        return [4 /*yield*/, (0, crypto_1.signWithSigningKey)(JSON.stringify(message), this.address)];
                    case 3:
                        signature = _f.sent();
                        return [4 /*yield*/, (0, crypto_1.getPublicKey)(this.address)];
                    case 4:
                        publicKey = _f.sent();
                        params = {
                            address: address,
                            handle: handle,
                            message: JSON.stringify(message),
                            signature: signature,
                            signingKey: publicKey,
                        };
                        return [4 /*yield*/, (0, queries_1.unfollow)(params, this.endpoint.cyberConnectApi)];
                    case 5:
                        resp = _f.sent();
                        if (!(((_a = resp === null || resp === void 0 ? void 0 : resp.data) === null || _a === void 0 ? void 0 : _a.unfollow.status) === 'INVALID_SIGNATURE')) return [3 /*break*/, 7];
                        return [4 /*yield*/, (0, crypto_1.clearSigningKey)()];
                    case 6:
                        _f.sent();
                        throw new error_1.ConnectError(error_1.ErrorCode.GraphqlError, (_b = resp === null || resp === void 0 ? void 0 : resp.data) === null || _b === void 0 ? void 0 : _b.unfollow.status);
                    case 7:
                        if (((_c = resp === null || resp === void 0 ? void 0 : resp.data) === null || _c === void 0 ? void 0 : _c.unfollow.status) !== 'SUCCESS') {
                            throw new error_1.ConnectError(error_1.ErrorCode.GraphqlError, (_d = resp === null || resp === void 0 ? void 0 : resp.data) === null || _d === void 0 ? void 0 : _d.unfollow.status);
                        }
                        return [3 /*break*/, 9];
                    case 8:
                        e_3 = _f.sent();
                        throw new error_1.ConnectError(error_1.ErrorCode.GraphqlError, e_3.message || e_3);
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    return CyberConnect;
}());
exports.default = CyberConnect;
